From 27609935b2755b9ec1aace408c46f5df7e591c66 Mon Sep 17 00:00:00 2001
From: "Pavel I. Kryukov" <kryukov@frtk.ru>
Date: Sun, 15 Oct 2017 11:46:43 +0300
Subject: [PATCH 06/12] Check GCC version instead of checking __AES__ macro

---
 sshaes.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/sshaes.c b/sshaes.c
index 7b42de8..5bdf38d 100644
--- a/sshaes.c
+++ b/sshaes.c
@@ -1126,7 +1126,7 @@ const struct ssh2_ciphers ssh2_aes = {
 #ifdef _FORCE_AES_NI
 #   define COMPILER_SUPPORTS_AES_NI
 #elif defined(__GNUC__)
-#    ifdef __AES__
+#    if ((__GNUC__ > 4) || (__GNUC__ == 4 && (__GNUC_MINOR__ >= 4))) && (defined(__x86_64__) || defined(__i386))
 #       define COMPILER_SUPPORTS_AES_NI
 #    endif
 #elif defined (_MSC_VER)
-- 
2.5.3.windows.1

