From 68a41432603217af24bea3c709d6330b3ec42e4f Mon Sep 17 00:00:00 2001
From: "Pavel I. Kryukov" <kryukov@frtk.ru>
Date: Sun, 15 Oct 2017 19:00:53 +0300
Subject: [PATCH 11/12] Check both AES and SSE4.1 extensions

---
 sshaes.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/sshaes.c b/sshaes.c
index bb6925e..5152bc6 100644
--- a/sshaes.c
+++ b/sshaes.c
@@ -1169,7 +1169,7 @@ INLINE static int supports_aes_ni()
 {
     unsigned int CPUInfo[4];
     __cpuid(1, CPUInfo[0], CPUInfo[1], CPUInfo[2], CPUInfo[3]);
-    return CPUInfo[2] & (1 << 25);
+    return (CPUInfo[2] & (1 << 25)) && (CPUInfo[2] & (1 << 19)); /* Check AES and SSE4.1 */
 }
 
 #else /* defined(__clang__) || defined(__GNUC__) */
@@ -1178,7 +1178,7 @@ INLINE static int supports_aes_ni()
 {
     unsigned int CPUInfo[4];
     __cpuid(CPUInfo, 1);
-    return CPUInfo[2] & (1 << 25);
+    return (CPUInfo[2] & (1 << 25)) && (CPUInfo[2] & (1 << 19)); /* Check AES and SSE4.1 */
 }
 
 #endif /* defined(__clang__) || defined(__GNUC__) */
-- 
2.5.3.windows.1

